### PR TITLE
Refactor lexer, parser and tests

### PR DESCRIPTION
**Pull request by xiam**
This refactor aims to improve debug/error messages (e.g.: error on line _x_ column _y_) and reduce complexity on RIDL's parser.

### CODE REVIEW

File: schema/ridl/_example/example1-partials.ridl [hunk 1]
```diff
@@ -68,13 +63,13 @@ service ExampleService
 
   - GetUser(req: GetUserRequest) => (user: User)
 
-  - Recv(req: stream string)
+  - stream Recv(req: string)
 
-  - Send() => (resp: stream string)
```
Comment (xiam) [inline schema/ridl/_example/example1-partials.ridl]: Support for `stream` arguments is removed (we'll be able to stream responses or requests but it would be difficult to stream single arguments).

File: schema/ridl/_example/example1.ridl [hunk 2]
```diff
@@ -0,0 +1,64 @@
+webrpc = v1
+
+name = hello-webrpc
+version = v0.0.1
+
+
+import
+  - ./example1-definitions.ridl
+
+import example1-partials.ridl
```
Comment (xiam) [inline schema/ridl/_example/example1.ridl]: Added support for partial imports.

File: schema/schema.go [hunk 3]
```diff
@@ -14,12 +14,17 @@ const (
 	VERSION = "v1"
 )
 
+type Import struct {
```
Comment (xiam) [inline schema/schema.go]: This change in the schema is required in order to support a full translation between the RIDL parser tree and the webrpc AST

File: schema/ridl/type_parser.go [hunk 4]
```diff
@@ -0,0 +1,228 @@
+package ridl
```
Comment (xiam) [inline schema/ridl/type_parser.go]: The type parser is still a bit hard to follow because of the `loop` jumps, we could get back to it later and refactor it with state functions (like the rest of the parser).

File: schema/ridl/parser_node.go [hunk 5]
```diff
@@ -0,0 +1,348 @@
+package ridl
```
Comment (xiam) [inline schema/ridl/parser_node.go]: This is an improved parser tree, it makes it easier to represent RIDL in terms of tokens (which will give us improved error messages) and to translate it to other formats, such as the WebRPC AST.

**Comment by xiam** [comment]
Hey @pkieltyka,

This lexer/parser refactor will make it a lot easier to implement future syntax improvements and to display error messages that are more aware of their context.

This PR also closes: https://github.com/webrpc/webrpc/issues/13

**Comment by pkieltyka** [comment]
@xiam I finally had a chance to do a review of this PR, sorry for the delay.

it looks friggen amazing! thank you for this amazing work.

**Comment by xiam** [comment]
Awesome @pkieltyka, thanks! What else do we need?

**Comment by pkieltyka** [comment]
@xiam I've tested it on a number of our internal projects that are quite large, and everything is working well after a few minor fixes I pushed to this branch.

There are two remaining issues we still need to resolve before we can merge+release this version:

1. Imports work great, but it needs to keep track of the path of a file so it can properly resolve imports of a tree. For example, say we have this example:

* myproject/user-api/user-api.ridl
* myproject/user-api/types.ridl
* myproject/article-api/article-api.ridl
* myproject/article-api/types.ridl
* myproject/proto/common.ridl
* myproject/proto/health.ridl

and say then `myproject/user-api/user-api.ridl` imports `./types.ridl` and also imports `../proto/common.ridl`

then say that `myproject/proto/common.ridl` imports `./health.ridl`

when you run webrpc-gen on `myproject/user-api/user-api.ridl`, it will fail because it can't resolve `./health.ridl` as it is trying to fetch the file from the path of where `myproject/user-api/user-api.ridl` is located, but it should be relative from the file that is making the import statement.

2. I've noticed when there are ridl errors in a file, the line number and column numbers are wrong and feel kinda all over the place, so its difficult for a user who made an error to find out where in fact their error is.

both of these are pretty small, but both important for usability

**Comment by xiam** [comment]
> Imports work great, but it needs to keep track of the path of a file so it can properly resolve imports of a tree. 

This is a good case, I pushed a WIP that should solve this (it's still unpolished, but should be enough for testing if your case gets solved). Could you give it a try and see?

And I think I've found the reason of mismatched line numbers, seems like we're skipping newlines without increasing the line counter... going to check that too.

**Comment by pkieltyka** [comment]
@xiam nice, the filepath resolution for import paths work great now.

I'm still seeing the line numbers to be off, perhaps newlines, and maybe comments throwing it off too?

**Comment by xiam** [comment]
@pkieltyka I found and fixed the problem with weird line numbers, should be better now. How does it look on your side?

**Comment by pkieltyka** [comment]
@xiam well done dude, it works great! thank you for this. It's going to mark the next big release of webrpc, what a major enhancement thank you.

I'm going to merge this now and will prepare release notes in a few days and do a version bump for the release.
